### ISSUE TITLE
State can't be used in two stages at once (even with `get_driver`)

### ISSUE THREAD

**Issue by concave-sphere** [bevyengine/bevy#1865]
## Bevy version

0.5

## Operating system & version

Debian GNU/Linux "buster"

## What you did

 1. Invoke `app.add_state(MyState::StateOne)`
 2. Create a stage `MyStage`
 3. Invoke `app.add_system_set(State::<MyState>::get_driver())`
 4. Add some systems to `MyStage` via `SystemSet::on_update(MyState::StateOne)`, etc.
 5. Add a system that switches between states.

## What you expected to happen

Systems that trigger on state enter and exit should run when the state changes, regardless of stage.

## What actually happened

Only state change systems in the `Update` stage are executed.

I know that the program hasn't completely locked because if I put state change systems into `Update`, they work fine.  It _does_ lock if I don't add `State::<MyState>::get_driver()`, so that's having an effect.

This text in the [Bevy Rustdoc](https://docs.rs/bevy/0.5.0/bevy/app/struct.AppBuilder.html#method.add_state) made me think this is supposed to work:

> If you need to use your state in a different stage, consider using Self::add_state_to_stage or manually adding State::get_driver to additional stages you need it in.

## Additional information

Sample code:
[test.tar.gz](https://github.com/bevyengine/bevy/files/6289058/test.tar.gz)

**Comment by alice-i-cecile** [comment]
Currently, States are owned by Stages, like all other run_criteria :( This should be fixed as part of the grand effort in #1375.

**Comment by Ixentus** [comment]
Instead of `app.add_system_set(State::<MyState>::get_driver())` you need `add_state_to_stage(MyStage, MyState::StateOne)`. Internally it does `get_driver()`, but adds it to the stage.

**Comment by concave-sphere** [comment]
> Instead of `app.add_system_set(State::<MyState>::get_driver())` you need `add_state_to_stage(MyStage, MyState::StateOne)`. Internally it does `get_driver()`, but adds it to the stage.

I tried replacing

```
.add_system_set_to_stage(MyStage, State::<MyState>::get_driver())
```

with

```
.add_state_to_stage(MyStage, MyState::StateOne)
```

The behavior I see is exactly the same: only the systems in the `Update` stage are executed on state transitions.

As a workaround, I've modified the actual code I was working on to just pull in the `State` resource and check for state transitions by hand (by storing the last-seen state).

**Comment by nicopap** [comment]
This is a major issue for me. I'd be happy to contribute a fix.

**Comment by cart** [comment]
Pulling in @TheRawMeatball (author of the current state system). I recommend discussing alternative designs with them.

**Comment by alice-i-cecile** [comment]
More discussion is roughly centralized in #2801. The existing run-criteria based states are deeply limited.

**Comment by alice-i-cecile** [comment]
Fixed by #7267.
